File: Populating Next Right.java
/*Time Complexity : O(N)
Space Complexity : O(N). Not optimal.
Did it run on LeetCode : Yes
Any Problems? : No.
*/
class Solution {
    public Node connect(Node root) {
        
        if(root==null) return root; // if root null, return root
        
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        
        while(!queue.isEmpty())
        {
            Node current = root;
            current = queue.poll();
            if(current.left!=null && current.right!=null)
            {
                current.left.next = current.right; // Making the left child point to Right child.
                if(current.next!=null)  // This is the case used to connect childs of two different branches. *Very Important case* 
                {
                    current.right.next = current.next.left; // In the given example. Connects 5->6.
                }
                
                queue.offer(current.left); // add left node to queue
                queue.offer(current.right); // add right node to queue.
            }
        }
       return root; 
    }
}

File: populating_next_right.js
//TC: O(n)
//SC: O(H)
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null) {
            return root;
        }
        dfs(root.left,root.right);
        return root;
        
        
    }
    public void dfs(Node left, Node right){
        
        if(left == null){
            return;
        }
        left.next = right;
        dfs(left.left,left.right);
        dfs(left.right,right.left);
        dfs(right.left,right.right);
    }
}